txnbuild/transaction.go

```diff
@@ -26,425 +28,742 @@ import (
-		tx.BaseFee = DefaultBaseFee
+func cloneEnvelope(e xdr.TransactionEnvelope, signatures []xdr.DecoratedSignature) (xdr.TransactionEnvelope, error) {
+	var clone xdr.TransactionEnvelope
+	binary, err := marshallBinary(e, signatures)
+	if err != nil {
+		return clone, errors.Wrap(err, "could not marshall envelope")
 	}
 
-	err := tx.setTransactionFee()
-	if err != nil {
-		panic(err)
+	if err = xdr.SafeUnmarshal(binary, &clone); err != nil {
+		return clone, errors.Wrap(err, "could not unmarshall envelope")
 	}
+	return clone, nil
 }
 
-// Build for Transaction completely configures the Transaction. After calling Build,
-// the Transaction is ready to be serialised or signed.
-func (tx *Transaction) Build() error {
-	// If transaction envelope has been signed, don't build transaction
-	if tx.xdrEnvelope != nil {
-		if len(tx.xdrEnvelope.Signatures) > 0 {
-			return errors.New("transaction has already been signed, so cannot be rebuilt.")
-		}
-		// clear the existing XDR so we don't append to any existing fields
-		tx.xdrEnvelope = &xdr.TransactionV0Envelope{}
-	}
+// Transaction represents a Stellar transaction. See
+// https://www.stellar.org/developers/guides/concepts/transactions.html
+// A Transaction may be wrapped by a FeeBumpTransaction in which case
+// the account authorizing the FeeBumpTransaction will pay for the transaction fees
+// instead of the Transaction's source account.
+type Transaction struct {
+	envelope      xdr.TransactionEnvelope
+	baseFee       int64
+	maxFee        int64
+	sourceAccount SimpleAccount
+	operations    []Operation
+	memo          Memo
+	timebounds    Timebounds
+	signatures    []xdr.DecoratedSignature
+}
+
+// BaseFee returns the per operation fee for this transaction.
+func (t *Transaction) BaseFee() int64 {
+	return t.baseFee
+}
 
-	// reset tx.xdrTransaction
-	tx.xdrTransaction = xdr.TransactionV0{}
+// MaxFee returns the total fees which can be spent to submit this transaction.
+func (t *Transaction) MaxFee() int64 {
+	return t.maxFee
+}
 
-	address := tx.SourceAccount.GetAccountID()
-	// Public keys start with 'G'
-	if address[0] != 'G' {
-		return errors.New("invalid public key for transaction source account")
-	}
-	accountID, err := xdr.AddressToAccountId(address)
+// SourceAccount returns the account which is originating this account.
+func (t *Transaction) SourceAccount() SimpleAccount {
+	return t.sourceAccount
+}
+
+// Memo returns the memo configured for this transaction.
+func (t *Transaction) Memo() Memo {
+	return t.memo
+}
+
+// Timebounds returns the Timebounds configured for this transaction.
+func (t *Transaction) Timebounds() Timebounds {
+	return t.timebounds
+}
+
+// Operations returns the list of operations included in this transaction.
+// The returned slice should not be modified.
+func (t *Transaction) Operations() []Operation {
+	return t.operations
+}
+
+// Signatures returns the list of signatures attached to this transaction.
+// The returned slice should not be modified.
```

Comment:
Nitpick (the slice itself is returned as value).

```suggestion
// The contents of the returned slice should not be modified.
```